src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -3644,16 +3644,41 @@ void addToStringMethod(MethodSymbol printMethod)
 
                 if (!memberSignatures.TryGetValue(targetMethod, out Symbol? existingToStringMethod))
                 {
-                    var toStringMethod = new SynthesizedRecordToString(this, printMethod, memberOffset: members.Count, diagnostics);
-                    members.Add(toStringMethod);
+                    var baseToStringMethod = getBaseToStringMethod();
+                    if (baseToStringMethod is null || !baseToStringMethod.IsSealed)
+                    {
+                        var toStringMethod = new SynthesizedRecordToString(this, printMethod, memberOffset: members.Count, diagnostics);
+                        members.Add(toStringMethod);
+                    }
                 }
                 else
                 {
                     var toStringMethod = (MethodSymbol)existingToStringMethod;
                     if (!SynthesizedRecordObjectMethod.VerifyOverridesMethodFromObject(toStringMethod, SpecialType.System_String, diagnostics) && toStringMethod.IsSealed && !IsSealed)
```

Comment:
I mean in cases when the override of Object.ToString is sealed up the hierarchy.

---
In reply to: [601507591](https://github.com/dotnet/roslyn/pull/52029#discussion_r601507591) [](ancestors = 601507591)

Reply:
@AlekseyTs

> If we allow sealing Object.ToString up the hierarchy, it feels like we should allow to declare new ToString method with the same parameter types that doesn't override Object.ToString, but shadows it instead.

Why is it currently forbidden in records anyway?
Anyway, wouldn't it be out of scope for this PR? I set out to allow sealing `ToString()`, not remove a restriction I didn't even know about...

Reply:
>> If we allow sealing Object.ToString up the hierarchy, it feels like we should allow to declare new ToString method with the same parameter types that doesn't override Object.ToString, but shadows it instead.
>> I mean in cases when the override of Object.ToString is sealed up the hierarchy.

>Why is it currently forbidden in records anyway?
>Anyway, wouldn't it be out of scope for this PR? I set out to allow sealing ToString(), not remove a restriction I didn't even know about...

This wasn't specifically forbidden. However, there was a requrement that a ```string ToString()``` method added explicitly overrides ```Object.ToString```. We are lifting this requirement for cases when ```Object.ToString``` is sealed in a base class. Therefore, when one adds ```string ToString()``` method in a situation like that, compiler shouldn't require that that method overrides ```Object.ToString```, because this simply cannot be done. It is not posible to override a sealed method. Please create a unit-test for this scenaro, I think that should give you the idea what I am talking about. 


---
In reply to: [602653210](https://github.com/dotnet/roslyn/pull/52029#discussion_r602653210) [](ancestors = 602653210)

Reply:
> Please create a unit-test for this scenaro, I think that should give you the idea what I am talking about.

@AlekseyTs do you mean something like this?

```csharp
        [Fact]
        public void ToString_NewToString_SealedBaseToString()
        {
            var source = @"
public record A
{
    public sealed override string ToString() => throw null;
}

public record B : A
{
    public new string ToString() => throw null;
}";

            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
            comp.VerifyEmitDiagnostics();
        }
```

This currently fails with this error:

```
// (9,23): error CS8869: 'B.ToString()' does not override expected method from 'object'.
//     public new string ToString() => throw null;
```

Should I make this test pass, then?

Reply:
> do you mean something like this?

Yes.

> Should I make this test pass, then?

I think yes.

---
In reply to: [602934953](https://github.com/dotnet/roslyn/pull/52029#discussion_r602934953) [](ancestors = 602934953)